cli: TestRemoveDeadReplicas failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/b775ca0318125c996305e4ab560e75d2d3471547

Parameters:
```
TAGS=
GOFLAGS=-race
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestRemoveDeadReplicas PKG=github.com/cockroachdb/cockroach/pkg/cli TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1037013&tab=buildLog

```
E181203 07:56:50.681354 5046 storage/replica.go:1521  [n1,s1,r10/4:/Table/1{3-4}] lease repl=(n2,s2):3 seq=3 start=1543823807.989575777,0 epo=1 pro=1543823807.989589450,0 owned by replica (n2,s2):3 that no longer exists
E181203 07:56:50.731240 5046 storage/replica.go:1521  [n1,s1,r10/4:/Table/1{3-4}] lease repl=(n2,s2):3 seq=3 start=1543823807.989575777,0 epo=1 pro=1543823807.989589450,0 owned by replica (n2,s2):3 that no longer exists
E181203 07:56:50.836290 5046 storage/replica.go:1521  [n1,s1,r10/4:/Table/1{3-4}] lease repl=(n2,s2):3 seq=3 start=1543823807.989575777,0 epo=1 pro=1543823807.989589450,0 owned by replica (n2,s2):3 that no longer exists
E181203 07:56:51.066039 5046 storage/replica.go:1521  [n1,s1,r10/4:/Table/1{3-4}] lease repl=(n2,s2):3 seq=3 start=1543823807.989575777,0 epo=1 pro=1543823807.989589450,0 owned by replica (n2,s2):3 that no longer exists
E181203 07:56:51.529526 5046 storage/replica.go:1521  [n1,s1,r10/4:/Table/1{3-4}] lease repl=(n2,s2):3 seq=3 start=1543823807.989575777,0 epo=1 pro=1543823807.989589450,0 owned by replica (n2,s2):3 that no longer exists
E181203 07:56:52.336206 5046 storage/replica.go:1521  [n1,s1,r10/4:/Table/1{3-4}] lease repl=(n2,s2):3 seq=3 start=1543823807.989575777,0 epo=1 pro=1543823807.989589450,0 owned by replica (n2,s2):3 that no longer exists
E181203 07:56:53.363270 5046 storage/replica.go:1521  [n1,s1,r10/4:/Table/1{3-4}] lease repl=(n2,s2):3 seq=3 start=1543823807.989575777,0 epo=1 pro=1543823807.989589450,0 owned by replica (n2,s2):3 that no longer exists
E181203 07:56:54.236658 5046 storage/replica.go:1521  [n1,s1,r10/4:/Table/1{3-4}] lease repl=(n2,s2):3 seq=3 start=1543823807.989575777,0 epo=1 pro=1543823807.989589450,0 owned by replica (n2,s2):3 that no longer exists
E181203 07:56:55.306877 5046 storage/replica.go:1521  [n1,s1,r10/4:/Table/1{3-4}] lease repl=(n2,s2):3 seq=3 start=1543823807.989575777,0 epo=1 pro=1543823807.989589450,0 owned by replica (n2,s2):3 that no longer exists
E181203 07:56:56.253737 5046 storage/replica.go:1521  [n1,s1,r10/4:/Table/1{3-4}] lease repl=(n2,s2):3 seq=3 start=1543823807.989575777,0 epo=1 pro=1543823807.989589450,0 owned by replica (n2,s2):3 that no longer exists
E181203 07:56:57.201677 5046 storage/replica.go:1521  [n1,s1,r10/4:/Table/1{3-4}] lease repl=(n2,s2):3 seq=3 start=1543823807.989575777,0 epo=1 pro=1543823807.989589450,0 owned by replica (n2,s2):3 that no longer exists
W181203 07:56:59.988268 4422 server/node.go:886  [n1,summaries] health alerts detected: {Alerts:[{StoreID:1 Category:METRICS Description:ranges.underreplicated Value:20} {StoreID:1 Category:METRICS Description:queue.replicate.process.failure Value:77}]}
W181203 07:57:02.950137 4409 storage/node_liveness.go:518  [n1,hb] slow heartbeat took 4.5s
W181203 07:57:02.950576 4409 storage/node_liveness.go:454  [n1,hb] failed node liveness heartbeat: context deadline exceeded
W181203 07:57:03.284522 4560 storage/node_liveness.go:518  [n4,hb] slow heartbeat took 4.3s
W181203 07:57:03.286584 4729 storage/node_liveness.go:518  [n5,hb] slow heartbeat took 3.8s
W181203 07:57:03.539064 5249 storage/raft_transport.go:584  [n4] while processing outgoing Raft queue to node 1: EOF:
W181203 07:57:03.541865 5317 storage/raft_transport.go:584  [n5] while processing outgoing Raft queue to node 1: EOF:
W181203 07:57:03.546901 5494 internal/client/txn.go:532  [n1,replicate,s1,r9/4:/Table/1{2-3}] failure aborting transaction: node unavailable; try another peer; abort caused by: log-range-event: node unavailable; try another peer
E181203 07:57:03.547399 5494 storage/queue.go:837  [n1,replicate,s1,r9/4:/Table/1{2-3}] change replicas of r9 failed: log-range-event: node unavailable; try another peer
W181203 07:57:03.553132 5122 storage/raft_transport.go:584  [n1] while processing outgoing Raft queue to node 5: EOF:
W181203 07:57:03.556046 5264 storage/raft_transport.go:584  [n1] while processing outgoing Raft queue to node 4: EOF:
W181203 07:57:03.566922 4722 gossip/gossip.go:1501  [n5] no incoming or outgoing connections
--- FAIL: TestRemoveDeadReplicas (22.49s)
	debug_test.go:247: expected replicas on {1,4,5} but got {1,4}

```

--- a/pkg/storage/store.go
+++ b/pkg/storage/store.go
@@ -4151,7 +4151,7 @@ func (s *Store) updateReplicationGauges(ctx context.Context) error {
 	if s.cfg.NodeLiveness != nil {
 		livenessMap = s.cfg.NodeLiveness.GetIsLiveMap()
 	}
-	availableNodes := s.cfg.StorePool.AvailableNodeCount()
+	availableNodes := s.AvailableNodeCount()
 
 	newStoreReplicaVisitor(s).Visit(func(rep *Replica) bool {
 		metrics := rep.Metrics(ctx, timestamp, livenessMap, availableNodes)
@@ -4315,6 +4315,17 @@ func (s *Store) ComputeMetrics(ctx context.Context, tick int) error {
 		}
 	}
 	return nil
+}
+
+// AvailableNodeCount returns this store's view of the number of nodes
+// in the cluster. This is the metric used for adapative zone configs;
+// ranges will not be reported as underreplicated if it is low. Tests
+// that wait for full replication by tracking the underreplicated
+// metric must also check for the expected AvailableNodeCount to avoid
+// catching the cluster while the first node is initialized but the
+// other nodes are not.
+func (s *Store) AvailableNodeCount() int {
+	return s.cfg.StorePool.AvailableNodeCount()
 }
 
 // StoreKeySpanStats carries the result of a stats computation over a key range.

--- a/pkg/testutils/testcluster/testcluster.go
+++ b/pkg/testutils/testcluster/testcluster.go
@@ -553,6 +553,11 @@ func (tc *TestCluster) WaitForFullReplication() error {
 		notReplicated = false
 		for _, s := range tc.Servers {
 			err := s.Stores().VisitStores(func(s *storage.Store) error {
+				if n := s.AvailableNodeCount(); n != len(tc.Servers) {
+					log.Infof(context.TODO(), "%s only sees %d/%d available nodes", s, n, len(tc.Servers))
+					notReplicated = true
+					return nil
+				}
 				if err := s.ComputeMetrics(context.TODO(), 0); err != nil {
 					// This can sometimes fail since ComputeMetrics calls
 					// updateReplicationGauges which needs the system config gossiped.

